{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f8e6c3f0",
   "metadata": {},
   "source": [
    "# Training classifiers to predict the presence of mutations in specific genes using bulk AML RNA-seq data, specifically Beat AML waves 1-4. These signatures will later be applied to the Van Galen single-cell RNA-seq data.\n",
    "\n",
    "Andrew Ashford, Pathways + Omics Group, Oregon Health & Science University - 11/14/2024\n",
    "\n",
    "This Jupyter Notebook will read in the previously-integrated AnnData object in which I used Scanorama to integrate both the Beat AML data and the Van Galen single-cell RNA-seq data. I used 200 lower-dimensional features for this integration. I will then apply the classifiers to the single-cell data and predict mutation status for several genes of interest that are associated with AML. Of note, all features were used that were shared between the Beat AML data and the Van Galen data (total = 22681 features). We will additionally apply the signatures to the single-cell data and then save the results and use them for downstream mutation prediction validation and analysis.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "686ec399",
   "metadata": {},
   "source": [
    "#### Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1f8beec5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Import modules\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "from sklearn.linear_model import RidgeClassifierCV\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.model_selection import cross_val_score, StratifiedKFold\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import accuracy_score\n",
    "from sklearn.utils import resample\n",
    "from sklearn.metrics import roc_auc_score\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ccb52ef",
   "metadata": {},
   "source": [
    "#### Specify directory and file locations and load in integrated data object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7d091e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the Van Galen directory location\n",
    "van_galen_data_dir = '/home/groups/precepts/ashforda/scOPE_github_stuff/data/testing/vanGalen_all_h5ad/'\n",
    "\n",
    "# Specify the Scanorama-integrated h5ad file name\n",
    "#integrated_data_file = 'Scanorama-integrated_combined_Beat_AML_Van_Galen_AML_adata_all_shared_features_22681.h5ad'\n",
    "#integrated_data_file = 'Scanorama-integrated_combined_Beat_AML_Van_Galen_AML_adata_all_shared_features_22681_imputed_scRNA.h5ad'\n",
    "integrated_data_file = 'Scanorama-integrated_combined_Beat_AML_Van_Galen_AML_adata_8000_scRNA_features_imputed_scRNA.h5ad'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0c01cd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the AnnData object\n",
    "integrated_data = sc.read_h5ad(van_galen_data_dir + integrated_data_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c039fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sanity check\n",
    "print(integrated_data)\n",
    "print(integrated_data.obs)\n",
    "print(integrated_data.X)\n",
    "print(integrated_data.X.max())\n",
    "print(integrated_data.X.min())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c0ee090",
   "metadata": {},
   "outputs": [],
   "source": [
    "# UMAP visualization of integrated data as a check\n",
    "sc.pp.neighbors(integrated_data)\n",
    "sc.tl.umap(integrated_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e8a2cfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Look at UMAP by cell type\n",
    "sc.pl.umap(integrated_data, color=['CellType'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb1fa449",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Look at UMAP by modality type\n",
    "sc.pl.umap(integrated_data, color=['batch'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f300a03",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "553cd0ed",
   "metadata": {},
   "source": [
    "#### Specify file locations and load in the Beat AML mutation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f92974a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the bulk RNA-seq mutation data\n",
    "mut_data = pd.read_csv('/home/groups/precepts/ashforda/scOPE_github_stuff/data/training/BeatAML/publicly_available_and_raw_counts/beataml_wes_wv1to4_mutations_dbgap.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e7b006",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make an object that spits the bulk data from the rest\n",
    "# Should be 707 samples.\n",
    "bulk_data = integrated_data[integrated_data.obs['batch'] == 'bulk']\n",
    "\n",
    "print(bulk_data)\n",
    "print(bulk_data.obs_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caeb716f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Need to add a different suffix to the sample names to match them with the mutation data\n",
    "bulk_data.obs_names = bulk_data.obs_names.str.replace('R-bulk', 'D')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b57c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modify the sample IDs to match between RNA and mutation data\n",
    "bulk_data.obs['dbgap_sample_id'] = bulk_data.obs.index.str.replace('R-1', 'D')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "897f00a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sanity check the mutation data - make sure the \"dbgap_sample_id\" column matches the format of the above\n",
    "# samples minus their \"R-bulk\" string replaced with \"D\"\n",
    "print(mut_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "876458cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensure the mutation labels are binary\n",
    "mutations = mut_data['symbol'].unique()\n",
    "\n",
    "print(mutations)\n",
    "print(len(mutations))\n",
    "print(set(mutations))\n",
    "print(len(set(mutations)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05711c2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove duplicate sample IDs in mutation data (if any)\n",
    "mut_data = mut_data.drop_duplicates(subset=['dbgap_sample_id', 'symbol'])\n",
    "\n",
    "# Create a binary matrix for mutation presence\n",
    "mut_data_binary = pd.crosstab(mut_data['dbgap_sample_id'], mut_data['symbol'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5291471",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(bulk_data.obs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8512109b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(mut_data_binary)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "780de012",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Align mutation data with bulk RNA-seq data\n",
    "bulk_sample_ids = bulk_data.obs['dbgap_sample_id']\n",
    "mut_data_agg = mut_data_binary.reindex(bulk_sample_ids, fill_value=0)\n",
    "\n",
    "print(mut_data_agg)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00f7ebc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(bulk_data.X)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "225cdb42",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0bb2509",
   "metadata": {},
   "source": [
    "#### Train classifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19882835",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Just need to set the \"X_pca_scaled variable\", the variable name doesn't make a ton of sense but it's what\n",
    "# the code downstream is using. It's just the values for the bulk RNA-seq data\n",
    "X_pca_scaled = bulk_data.X\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb61acea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize models\n",
    "ridge_model = RidgeClassifierCV()\n",
    "rf_model = RandomForestClassifier(n_estimators=100, random_state=42)\n",
    "svm_model = SVC(kernel='linear', probability=True, random_state=42)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84ab390",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to train and evaluate model with Stratified K-Folds\n",
    "def train_evaluate_model(X, y, model, model_name, cv_splits, mut_name, auc_dict):\n",
    "    skf = StratifiedKFold(n_splits=cv_splits, shuffle=True, random_state=42)\n",
    "    #scores = cross_val_score(model, X, y, cv=skf, scoring='accuracy')\n",
    "    scores = cross_val_score(model, X, y, cv=skf, scoring='roc_auc')\n",
    "    #print(f\"{model_name} - Cross-validation accuracy: {np.mean(scores)}\")\n",
    "    \n",
    "    model_rstrip = model_name.rstrip(' ')\n",
    "    auc_dict[mut_name][model_rstrip].append(np.mean(scores))\n",
    "    \n",
    "    print(f\"{model_name} - Cross-validation ROC AUC: {np.mean(scores)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c37d57c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# User-defined minimum number of positive samples required\n",
    "MIN_POSITIVE_SAMPLES = 10\n",
    "DEFAULT_CV_SPLITS = 10\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df28c02a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Track mutations with sufficient positive samples\n",
    "valid_mutations = []\n",
    "\n",
    "# Keep the ROC AUC scores for each mutation in a dictionary\n",
    "roc_auc_dict = {}\n",
    "\n",
    "# Train and evaluate models\n",
    "for mutation in mutations:\n",
    "    #print(mutation)\n",
    "    if mutation in mut_data_agg.columns:\n",
    "        #print('mutation is in mut_data_agg.columns')\n",
    "        \n",
    "        y_mutation = mut_data_agg[mutation].astype(int)\n",
    "        #print(y_mutation)\n",
    "        \n",
    "        # Check the distribution of the labels to ensure it makes sense\n",
    "        positive_samples = np.sum(y_mutation)\n",
    "        negative_samples = len(y_mutation) - positive_samples\n",
    "        \n",
    "        if positive_samples < MIN_POSITIVE_SAMPLES:\n",
    "            continue\n",
    "        \n",
    "        # Dynamically set the number of CV splits\n",
    "        cv_splits = min(DEFAULT_CV_SPLITS, positive_samples, negative_samples)\n",
    "        \n",
    "        if cv_splits < 2:\n",
    "            continue\n",
    "            \n",
    "        if mutation not in roc_auc_dict.keys():\n",
    "            roc_auc_dict[mutation] = {'Ridge Classifier': [], 'Random Forest Classifier': [], 'SVM Classifier': []}\n",
    "        \n",
    "        #print(roc_auc_dict)\n",
    "        \n",
    "        print(f\"Training models for mutation: {mutation}\")\n",
    "        \n",
    "        # Sample equal number of negative samples\n",
    "        positive_indices = np.where(y_mutation == 1)[0]\n",
    "        negative_indices = np.where(y_mutation == 0)[0]\n",
    "        \n",
    "        negative_indices_sampled = resample(negative_indices, replace=False, n_samples=positive_samples, random_state=42)\n",
    "        \n",
    "        # Combine positive and negative samples\n",
    "        combined_indices = np.concatenate([positive_indices, negative_indices_sampled])\n",
    "        X_pca_scaled_sampled = X_pca_scaled[combined_indices]\n",
    "        y_mutation_sampled = y_mutation.iloc[combined_indices]\n",
    "        \n",
    "        train_evaluate_model(X_pca_scaled_sampled, y_mutation_sampled, ridge_model, \"Ridge Classifier \", cv_splits, mutation, roc_auc_dict)\n",
    "        train_evaluate_model(X_pca_scaled_sampled, y_mutation_sampled, rf_model, \"Random Forest Classifier \", cv_splits, mutation, roc_auc_dict)\n",
    "        train_evaluate_model(X_pca_scaled_sampled, y_mutation_sampled, svm_model, \"SVM Classifier \", cv_splits, mutation, roc_auc_dict)\n",
    "        \n",
    "        valid_mutations.append(mutation)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90dc88ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c996d54",
   "metadata": {},
   "source": [
    "#### Validate some of the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5db17869",
   "metadata": {},
   "outputs": [],
   "source": [
    "# AUC score dict sanity check\n",
    "print(roc_auc_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f334784",
   "metadata": {},
   "outputs": [],
   "source": [
    "performance_dict = roc_auc_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79fbd153",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "# Assuming performance_dict is your dictionary with the classifier performances\n",
    "\n",
    "# Extract unique classifier names\n",
    "classifiers = list(next(iter(performance_dict.values())).keys())\n",
    "\n",
    "# Prepare data for plotting\n",
    "genes = list(performance_dict.keys())\n",
    "x = np.arange(len(genes))  # Gene indices\n",
    "width = 0.2  # Width of the bars\n",
    "\n",
    "# Initialize figure and axis\n",
    "plt.figure(figsize=(20, 10))\n",
    "\n",
    "# Create a bar for each classifier\n",
    "for i, clf_name in enumerate(classifiers):\n",
    "    means = [np.mean(performance_dict[gene][clf_name]) for gene in genes]\n",
    "    stds = [np.std(performance_dict[gene][clf_name]) for gene in genes]\n",
    "    plt.bar(x + i*width, means, width, yerr=stds, label=clf_name, capsize=5)\n",
    "\n",
    "# Customize the plot\n",
    "plt.xticks(x + width, genes, rotation=90)\n",
    "plt.xlabel('Genes')\n",
    "plt.ylabel('Performance (AUC)')\n",
    "plt.title('Classifier Performance Across Genes')\n",
    "plt.ylim(0, 1)\n",
    "plt.legend()\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1982ce98",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "becf4abd",
   "metadata": {},
   "source": [
    "#### Apply trained mutation classifiers to the single-cell data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a249ac39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Again, this variable name doesn't make a ton of sense, but it's just scRNA AnnData .X values\n",
    "#scRNA_data = integrated_data[integrated_data.obs['dataset'] == '0']\n",
    "scRNA_data = integrated_data[integrated_data.obs['batch'] == 'scRNA']\n",
    "\n",
    "# NOTE: Using .X instead of .obsm['X_pca']\n",
    "#X_scRNA_pca = scRNA_data.obsm['X_pca']\n",
    "X_scRNA_pca = scRNA_data.X\n",
    "#X_scRNA_pca_scaled = scaler.transform(X_scRNA_pca)\n",
    "X_scRNA_pca_scaled = X_scRNA_pca\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2abead6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apply trained models to scRNA-seq data\n",
    "print('Beginning to apply trained models to scRNA-seq data..')\n",
    "\n",
    "# Dictionary to store prediction results\n",
    "predictions = {}\n",
    "\n",
    "print(valid_mutations)\n",
    "\n",
    "# Apply predictions only for valid mutations\n",
    "for mutation in valid_mutations:\n",
    "    \n",
    "    print(f'Predicting mutation status in single-cell data for mutation: {mutation}..')\n",
    "    y_mutation = mut_data_agg[mutation].astype(int)\n",
    "    \n",
    "    # Train the final models\n",
    "    ridge_model.fit(X_pca_scaled, y_mutation)\n",
    "    rf_model.fit(X_pca_scaled, y_mutation)\n",
    "    svm_model.fit(X_pca_scaled, y_mutation)\n",
    "    \n",
    "    # Predict using scRNA-seq data\n",
    "    scRNA_data.obs[f'ridge_preds_{mutation}'] = ridge_model.predict(X_scRNA_pca_scaled)\n",
    "    scRNA_data.obs[f'rf_preds_{mutation}'] = rf_model.predict(X_scRNA_pca_scaled)\n",
    "    scRNA_data.obs[f'svm_preds_{mutation}'] = svm_model.predict(X_scRNA_pca_scaled)\n",
    "    \n",
    "    # Save predictions to dictionary\n",
    "    predictions[f'ridge_preds_{mutation}'] = scRNA_data.obs[f'ridge_preds_{mutation}']\n",
    "    predictions[f'rf_preds_{mutation}'] = scRNA_data.obs[f'rf_preds_{mutation}']\n",
    "    predictions[f'svm_preds_{mutation}'] = scRNA_data.obs[f'svm_preds_{mutation}']\n",
    "\n",
    "print(\"Models trained and predictions applied to scRNA-seq data.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e463a3ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Again, there are variables specified above for each of the datasets that have the variable names:\n",
    "# X_pca_scaled for the training features of the bulk data\n",
    "# X_scRNA_pca_scaled which are the features to use for the scRNA-seq data\n",
    "\n",
    "# If those were'nt specified above you can do so here:\n",
    "#X_pca = bulk_data.X\n",
    "#X_scRNA_pca_scaled = scRNA_data.X \n",
    "\n",
    "# These variables are used by the model to use the bulk features to train models, then apply the single-cell\n",
    "# features to get single-cell mutation classification probabilty predictions\n",
    "\n",
    "# Alternatively, if you wanted to train and predict using just the RNA features alone unintegrated, you could \n",
    "# set X_pca_scaled to the shared RNA features of the bulk data and X_scRNA_pca_scaled to the shared RNA features\n",
    "# of the scRNA data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09e28fa9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Can apply different settings to the ridge regression classifier to account for scRNA dropout issues\n",
    "'''\n",
    "import numpy as np\n",
    "from sklearn.linear_model import RidgeClassifierCV\n",
    "from sklearn.calibration import CalibratedClassifierCV\n",
    "\n",
    "# Define a custom range of alpha values\n",
    "alphas = np.logspace(-6, 6, 13)\n",
    "\n",
    "# Initialize the Ridge Classifier with the custom alpha values and balanced class weights\n",
    "ridge_classifier = RidgeClassifierCV(alphas=alphas, class_weight='balanced')\n",
    "\n",
    "# Wrap with CalibratedClassifierCV\n",
    "calibrated_ridge = CalibratedClassifierCV(ridge_classifier, method='sigmoid')\n",
    "\n",
    "# Fit the model\n",
    "calibrated_ridge.fit(X_train, y_train)\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e403bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alternatively, I used this code previously to get the PROBABILITY of having a mutation from the classifier.\n",
    "# This code does both the training and application to the single-cell data\n",
    "from sklearn.linear_model import RidgeClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.svm import SVC\n",
    "\n",
    "# Modules for new models\n",
    "from sklearn.calibration import CalibratedClassifierCV\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier\n",
    "\n",
    "# Original models\n",
    "'''\n",
    "# Initialize your models\n",
    "#ridge_model = RidgeClassifier()  # Note: RidgeClassifier does not have predict_proba\n",
    "#rf_model = RandomForestClassifier()\n",
    "#svm_model = SVC(probability=True)  # Make sure SVM is configured to output probabilities\n",
    "'''\n",
    "# Model training with hyperparameter tuning and calibration\n",
    "ridge_model = CalibratedClassifierCV(RidgeClassifierCV(), method='sigmoid')\n",
    "rf_model = GridSearchCV(RandomForestClassifier(), param_grid={'n_estimators': [100, 200], 'max_depth': [5, 10]}, cv=5)\n",
    "svm_model = CalibratedClassifierCV(SVC(probability=True), method='sigmoid')\n",
    "#gbm_model = GridSearchCV(GradientBoostingClassifier(), param_grid={'n_estimators': [100, 200], 'learning_rate': [0.01, 0.1]}, cv=5)\n",
    "\n",
    "# Apply trained models to scRNA-seq data\n",
    "print('Beginning to apply trained models to scRNA-seq data..')\n",
    "\n",
    "# Dictionary to store prediction results\n",
    "predictions = {}\n",
    "\n",
    "print(valid_mutations)\n",
    "\n",
    "# Apply predictions only for valid mutations\n",
    "for mutation in valid_mutations:\n",
    "    print(f'Predicting mutation status in single-cell data for mutation: {mutation}..')\n",
    "    y_mutation = mut_data_agg[mutation].astype(int)\n",
    "    \n",
    "    # Train the final models on bulk RNA-seq data\n",
    "    ridge_model.fit(X_pca_scaled, y_mutation)\n",
    "    rf_model.fit(X_pca_scaled, y_mutation)\n",
    "    svm_model.fit(X_pca_scaled, y_mutation)\n",
    "    #gbm_model.fit(X_pca_scaled, y_mutation)\n",
    "    \n",
    "    '''\n",
    "    # Predict using scRNA-seq data\n",
    "    # RidgeClassifier does not have predict_proba, using decision_function instead\n",
    "    ridge_probs = ridge_model.decision_function(X_scRNA_pca_scaled)\n",
    "    rf_probs = rf_model.predict_proba(X_scRNA_pca_scaled)[:, 1]\n",
    "    svm_probs = svm_model.predict_proba(X_scRNA_pca_scaled)[:, 1]\n",
    "    '''\n",
    "    # Predict using scRNA-seq data\n",
    "    ridge_probs = ridge_model.predict_proba(X_scRNA_pca_scaled)[:, 1]\n",
    "    rf_probs = rf_model.predict_proba(X_scRNA_pca_scaled)[:, 1]\n",
    "    svm_probs = svm_model.predict_proba(X_scRNA_pca_scaled)[:, 1]\n",
    "    #gbm_probs = gbm_model.predict_proba(X_scRNA_pca_scaled)[:, 1]\n",
    "    \n",
    "    # Combine predictions using an ensemble method (e.g., average)\n",
    "    # Idk if this is a common practice. Commented it out for now.\n",
    "    #ensemble_probs = (ridge_probs + rf_probs + svm_probs + gbm_probs) / 4\n",
    "    \n",
    "    scRNA_data.obs[f'ridge_preds_{mutation}'] = ridge_probs\n",
    "    scRNA_data.obs[f'rf_preds_{mutation}'] = rf_probs\n",
    "    scRNA_data.obs[f'svm_preds_{mutation}'] = svm_probs\n",
    "    # New preds\n",
    "    #scRNA_data.obs[f'gbm_preds_{mutation}'] = gbm_probs\n",
    "    #scRNA_data.obs[f'ensemble_preds_{mutation}'] = ensemble_probs\n",
    "    \n",
    "    # Save predictions to dictionary\n",
    "    predictions[f'ridge_preds_{mutation}'] = ridge_probs\n",
    "    predictions[f'rf_preds_{mutation}'] = rf_probs\n",
    "    predictions[f'svm_preds_{mutation}'] = svm_probs\n",
    "    # New preds\n",
    "    #predictions[f'gbm_preds_{mutation}'] = gbm_probs\n",
    "    #predictions[f'ensemble_preds_{mutation}'] = ensemble_probs\n",
    "\n",
    "    '''\n",
    "    # Additionally, save the predicted probabilities for the bulk RNA-seq data\n",
    "    ridge_bulk_probs = ridge_model.decision_function(X_pca_scaled)\n",
    "    rf_bulk_probs = rf_model.predict_proba(X_pca_scaled)[:, 1]\n",
    "    svm_bulk_probs = svm_model.predict_proba(X_pca_scaled)[:, 1]\n",
    "    '''\n",
    "    ridge_bulk_probs = ridge_model.predict_proba(X_pca_scaled)[:, 1]\n",
    "    rf_bulk_probs = rf_model.predict_proba(X_pca_scaled)[:, 1]\n",
    "    svm_bulk_probs = svm_model.predict_proba(X_pca_scaled)[:, 1]\n",
    "    \n",
    "    predictions[f'ridge_bulk_probs_{mutation}'] = ridge_bulk_probs\n",
    "    predictions[f'rf_bulk_probs_{mutation}'] = rf_bulk_probs\n",
    "    predictions[f'svm_bulk_probs_{mutation}'] = svm_bulk_probs\n",
    "\n",
    "print(\"Models trained and predictions applied to scRNA-seq data.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e233a0e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# The scRNA AnnData object now how the prediction scores within the .obs variable named after the classifier\n",
    "# used, then named after the gene it's predicting mutations in\n",
    "print(scRNA_data)\n",
    "print(scRNA_data.obs['ridge_preds_DNMT3A'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d037d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the scRNA predictions object for future analysis and validation\n",
    "scRNA_data.write(van_galen_data_dir + 'Scanorama-integrated_Beat_AML_Van_Galen_AML_adata_8000_scRNA_features_imputed_scRNA_single-cell_predictions.h5ad')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7635961a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scOPE_working_environment",
   "language": "python",
   "name": "scope_working_environment"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
